Add support for using sfpi from a system path

blozano-tt opened the pull request:
### Problem description
The current scheme of using TT_METAL_HOME to find the sfpi compiler causes some difficulty in shipping a package.
Today, we have to bundle the sfpi compiler along with ttnn wheel.
The wheel then becomes too large to distribute on pypi, as it becomes ~180MB.
The wheel has some "tricks" to automatically set TT_METAL_HOME and symlink to site-packages.
All this gets easier if we have an installer for sfpi, to install it to a known system path.

### What's changed
If the sfpi compiler does not exist in `$TT_METAL_HOME/runtime/` fallback to `/opt/tenstorrent/`.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/14393337744) CI passes
- [x] [All post commit after refactor](https://github.com/tenstorrent/tt-metal/actions/runs/14413097598) CI passes
- [x] New/Existing tests provide coverage for changes

blozano-tt committed:
Add support for system install of sfpi

blozano-tt committed:
Forward progress

blozano-tt committed:
Revert one change

blozano-tt commented:
I noticed the new prints I added are a little noisy when running unit tests.... should I switch to log_debug?

nathan-TT commented on tt_metal/jit_build/build.hpp:
```diff
@@ -79,7 +79,8 @@ class JitBuildEnv {
 
     // Tools
-    string gpp_;
+    string gpp_ = "";
```
Just above the gxx locating code is whether to use ccache, may as well remove the setting to `""` in the else branch there.  Oh, and IMHO the `// Tools` comment belongs just before choosing ccache.  care to make those small cleanups while you're here.

nathan-TT commented on tt_metal/jit_build/build.cpp:
```diff
@@ -134,7 +134,19 @@ void JitBuildEnv::init(
+    std::string system_gpp_path = "/opt/tenstorrent/sfpi/compiler/bin/riscv32-unknown-elf-g++";
+    if (std::filesystem::exists(local_gpp_path)) {
+        this->gpp_ += local_gpp_path + " ";
+        this->gpp_include_dir_ = this->root_ + "runtime/sfpi/include";
```
this is repeats the root when adding the include and (almost) repeats the log information.  How about something more like:
```
std::string const sfpi_roots[2] = {this->root_ + "runtime/sfpi", "/opt/tenstorrent/sfpi"};
for (unsigned ix = 0; ix != 2; ix++) {
  auto gxx = sfpi_roots[ix] + "/compiler/bin/riscv32-unknown-elf-g++";
  if (std::filesystem::exists(gxx)) {
    this->app_ += gxx;
    this->gpp_include_dir_ = sfpi_roots[ix] + "/include";
    log_info(tt::LogBuildKernels, "Using {} sfpi at {}", ix ? "system" : "local", sfpi_roots[ix]);
    goto found; // if you dislike this, test whether gpp_include_dir_'s been set after the loop.
}
TT_THROW("sfpi not found at {}", sfpi_roots); // does this print arrays?
found:;

```

blozano-tt replied:
Alright, I incorporated your suggestions in the latest commit. I like that your approach is more compact, and could be extended with more roots. I wasn't sure if fmt can handle printing arrays, so just left that unoptimized.

blozano-tt committed:
Update build.cpp

blozano-tt committed:
missing closing brace

nathan-TT reviewed (approved):
LGTM

blozano-tt committed:
Fix unexplainable bug

Without the clearing of gpp_ the path to the compiler is doubled...

i.e. `g++ g++`
